### ISSUE TITLE
Insert performance question

### ISSUE THREAD

**Issue by Daniel-Valliere** [orientechnologies/orientdb#4793]
Hi Luca,

I’ve tried to get the most performance for inserting a number of records and edges into OrientDB.
The best I got was the following. Do you have any idea how I could get better?

Create one vertex: 0.35 ms
Create one edge: 1.6 ms

In my setup, java code is in the same VM as OrientDB (embedded DB) and the DB is on an SSD drive.
I want to create 1000 instances of a vertex B and then for each instance create an edge to an instance of A
I’ve tried with SQL insert, SQL batch insert and straight Tinkerpop addVertex/createEdge. No performance differences.
I’ve tried with multiple worker thread in the same process but it appear to throttle down to single thread. Single thread gave better performance.
I’ve tried with  commit after any number of creation. Best performance is for a commit after each creation (vertex or edge)

So in summary, I’ve got the best insert performance by using a single thread to create the vertices and edges and doing a commit after each creation. Declaring the intent OIntentMassiveInsert makes no significant improvement (application has a hook but does nothing in it).

Is there something that can be done to get better performance?

**Comment by lvca** [comment]
The fastest way is avoiding SQL at all, but rather use the Java API.

**Comment by Daniel-Valliere** [comment]
public static void testBigInsert(int offset, int size) {
      OrientBaseGraph graph = new OrientGraph("plocal:/dev/tests/orientrepl/databases/test2");
//    graph.declareIntent(new OIntentMassiveInsert());
//    OGlobalConfiguration.TX_USE_LOG.setValue(false);

      try {
         OrientVertex[] devices = new OrientVertex[size];
         OrientVertex[] ports = new OrientVertex[size];
         
         long start;
         long end;
   
         // Create devices
         start = System.currentTimeMillis();
         for (int i = 0; i < size; i++) {
            int idx = offset + i;
            devices[i] = graph.addVertex("class:Device", "name", "N" + idx + "Device", "number", idx);
            devices[i].getProperty("name");
         }
         end = System.currentTimeMillis();
         System.out.println("Create devices: " + (end - start));
         
         // Create ports
         start = System.currentTimeMillis();
         for (int i = 0; i < size; i++) {
            int idx = offset + i;
            ports[i] = graph.addVertex("class:Port", "name", "N" + idx + "Port", "number", idx);
            ports[i].getProperty("name");
         }
         end = System.currentTimeMillis();
         System.out.println("Create ports: " + (end - start));
         
         // Create edges
         start = System.currentTimeMillis();
         for (int i = 0; i < devices.length; i++) {
            graph.addEdge(null, devices[i], ports[i], "portE");
         }
         graph.commit();
         end = System.currentTimeMillis();
         System.out.println("Create edges: " + (end - start));
//       graph.declareIntent(null);
      } catch (Throwable e) {
         log.error("Exception thrown: ", e);
      } finally {
         log.info("Shutting down");
         graph.shutdown();
      }
   }

**Comment by luigidellaquila** [comment]
Hi @Daniel-Valliere 

a multi-thread approach should give good results in this case, because you do not have collisions (multiple threads that try to add edges to the same vertex).
The only thing that can lead to lock contention in multi-thread in this case is an index update.
Do you have indexes on vertices and/or edges?

As a quick improvement, you can try to create two separate threads, one to create devices and one for ports.
After these threads have finished you can continue with edges.
This should give you about 10-15% speedup (if you do not have indices on V class or on a common superclass of Port and Device).

If you had a sample empty db (schema only) or a script to create your schema, I could try it on my side and try to optimize it for you

Thanks

Luigi

**Comment by Daniel-Valliere** [comment]
Thanks Luigi

I did try to create devices on one thread and ports on another thread and there is indeed speed improvement.
Once the devices + ports are created, I create the edges also using 2 threads and there is also a speed improvement there (all edges have no in or out vertices in common). The higher the number of devices/ports created, the more visible the improvement. 

From google search and looking at documentation, I've done the following. Do you have further suggestions for best performance?

Graph created in the following way.
  OrientGraphNoTx graph = new OrientGraphNoTx("plocal:/dev/tests/orientrepl/databases/test2"); // on SSD drive
  graph.getRawGraph().getLocalCache().setEnable(false);
  graph.declareIntent(new OIntentMassiveInsert());
  return graph;

JVM arguments on a (host has 16 gig of memory)
-server -XX:+AggressiveOpts -XX:CompileThreshold=200 -Xmx2048m -Dstorage.diskCache.bufferSize=7200

Our current results in the best scenario: a vertex is created in 0.016 ms and an edge is created in 0.083 ms. that gives use about 10,000 vertex + edge creation per second.

**Comment by luigidellaquila** [comment]
Hi @Daniel-Valliere 

it all seems correct. You can just try another alternative, but I guess the results will be similar: create multiple threads (as the number of your machine cores) and let each of them create triples vertex-edge-vertex in parallel, where each triple creation is made in a single transaction.

Another suggestion is to create multiple clusters per class, it will give you less lock contention on the storage with multithreading operations (see MINIMUMCLUSTERS http://orientdb.com/docs/last/SQL-Alter-Database.html and ADD CLUSTER http://orientdb.com/docs/last/SQL-Alter-Class.html)

**Comment by Daniel-Valliere** [comment]
Thanks Luigi,

Using transactional graph is slower than no transactional graph. So unless I misunderstood, creating the tripplet didn't help.

Indeed adding clusters to a class allows to create more ports and edges concurrently. Thanks. 

If there are no more suggestions for performance improvement, we can close this ticket. If there are, I'm all ears :)

Cheers

**Comment by luigidellaquila** [comment]
Hi @Daniel-Valliere 

Ok, thanks, closing.
Right now I do not have any other suggestions to share. I'll reach out if something else comes to my mind 

Luigi

**Comment by martingg88** [comment]
Do you think the following case that show the poor performance?
The following measure it running with java executor threading mechanism and graph transaction mode. 

It take about 1.5 hour to insert 20k vertex. 

each vertex will have following edge relationship

one user vertex
one activity vertex
one place vertex
around 10 to 20 photos vertex.
